src/loaders/genome_collection/parse_computed_genome_attribs.py

```diff
@@ -191,6 +234,233 @@ def checkm2(root_dir, kbase_collection, load_ver):
+            categories[trait_category] = ColumnCategory(category=trait_category, columns=[])
+
+        # add the trait to the appropriate category's list of columns
+        categories[trait_category].columns.append(ColumnInformation(id=str(trait[_SYS_TRAIT_INDEX]),
+                                                                    name=trait[_SYS_TRAIT_NAME],
+                                                                    description=trait[_SYS_TRAIT_DESCRIPTION],
+                                                                    type=ColumnType.COUNT.value))
+
+    return list(categories.values())
+
+
+def _is_float_int(num: Any) -> bool:
+    # Check if a number is an integer or a float with an integer value
+    return isinstance(num, int) or (isinstance(num, float) and num.is_integer())
+
+
+def _append_cell(
+        heatmap_row: HeatMapRow,
+        trait_idx: int,
+        cell_count: int,
+        min_value: float | int,
+        max_value: float | int,
+        trait_val: float | int | bool = 0) -> (float | int, float | int):
+    # Append a cell to the heatmap row and return the global min and max values
+
+    cell = Cell(celid=str(cell_count), colid=str(trait_idx), val=trait_val)
+    heatmap_row.cells.append(cell)
+
+    return min(min_value, trait_val), max(max_value, trait_val)
+
+
+def _parse_rows(
+        traits_meta: Dict[str, Dict[str, Union[int, str]]],
+        traits_val: Dict[str, List[Dict[str, Union[int, float, bool]]]],
+        ensure_ints: bool = False) -> (List[HeatMapRow], Union[float, int], Union[float, int]):
+    min_value, max_value, cell_count = float('inf'), float('-inf'), 0
+    heatmap_rows, trait_idxs = [], set([trait[_SYS_TRAIT_INDEX] for trait in traits_meta.values()])
+    for data_id, traits_val_list in traits_val.items():
+        heatmap_row = HeatMapRow(kbase_id=data_id, cells=[])
+        visited_traits = set()
+        for trait_val_info in traits_val_list:
+            trait_idx, trait_val = trait_val_info[_SYS_TRAIT_INDEX], trait_val_info[_SYS_TRAIT_VALUE]
+            visited_traits.add(trait_idx)
+            if ensure_ints:
+                if not _is_float_int(trait_val):
+                    raise ValueError(f'Trait value {trait_val} does not hold an integer value')
+                trait_val = int(trait_val)
+            min_value, max_value = _append_cell(heatmap_row, trait_idx, cell_count, min_value, max_value,
+                                                trait_val=trait_val)
+            cell_count += 1
+
+        # fill in missing trait values with 0s for all cells
+        missing_trait_idxs = trait_idxs - visited_traits
+        for missing_trait_idx in missing_trait_idxs:
+            min_value, max_value = _append_cell(heatmap_row, missing_trait_idx, cell_count, min_value, max_value,
+                                                trait_val=_SYS_DEFAULT_TRAIT_VALUE)
+            cell_count += 1
+
+        heatmap_rows.append(heatmap_row)
```

Comment:
This means that if different rows have different missing traits, the rows will have different cell orderings that are out of order with the meta columns, IIUC

Reply:
for cells, I think that's fine. Cells are mapped to columns via `colid`. But meta columns are not sorted either, right? Columns are grouped into categories which breaks ordering.

Reply:
I think this is an old comment, right? We're now enforcing the ordering from microtrait everywhere

Reply:
not at cells level. We are enforcing the column order. Cells has an reference to columns, I don't know why we need to enforce cell order.

Reply:
To make it easier for the front end to render it

Reply:
make it easier so that the rendering code does not enforce colid matches the actual column?

Reply:
So it doesn't have to, it could just check the order is ok rather than having to reorder the cells

Reply:
👍

Reply:
:+1: